Organs now freeze with atmosphere and core temperature

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

- Organs are frozen by 0 celsius atmosphere
- Organs are frozen in human mobs with core temperatures beneath 0 celsius AND beneath their species' cold damage limit
- Frozen organs in living mobs are now failing organs and will kill their owner if the owner doesn't warm up
- Frozen organs now look visually frozen and prepend the word 'frozen' to their names, like other frozen objects
- Human mobs with skin (not core) temperatures beneath their species' cold damage limit now look paler
- Human mobs frozen by morgue trays, freezer crates and freezer closets now look paler, to indicate their organs are frozen
- Can't do surgeries with organ_to_manipulate set if that organ is frozen
- Can't do surgeries if the patient is frozen (i.e. core temperature beneath both 0 celsius and their species' cold damage limit)

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

- Organs and bodies freezing in cold atmosphere creates more opportunities to preserve them, useful with an overrun medbay
- Making mobs pale when they're taking cold damage indicates it to other players
- Making mobs pale when their organs have been frozen by a container makes organ preservation in bodies more obvious 
- Making organs visually frozen makes organ preservation inside and outside of bodies obvious instead of having no easy way of knowing it's occurring
- Preventing surgeries on frozen mobs and frozen organs prevents using atmosphere to get around organ decay on dead mobs on operating tables which would be too easy

Todo
- [ ] Make morgue trays, freezer closets and freezer crates lower mobs' core temperatures to eventually freeze their organs, instead of doing it after a short delay and without regard for core temperature
- [ ] Unfreeze organs when entering cryo to prevent further organ damage to them, given cryo doesn't freeze them and who really wants to make cryo shittier
- [ ] Test temperature guns and cryokinesis
- [ ] Consider balancing advanced surgery on mobs with frozen organs (i.e. non-decaying organs on an operating table in a cold environment) using the slowdown cryostylane uses instead of preventing it entirely

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
balance: Organs now freeze in 0 celsius atmosphere
balance: Organs in human mobs freeze when the mob's core temperature is beneath 0 celsius AND their species' cold damage limit
balance: Frozen organs in living mobs are failing organs and will kill their owner if they don't warm up
add: Frozen organs are now visibly frozen
add: Human mobs now look paler when their skin temperature is beneath their species' cold damage limit
add: Human mobs now look paler when their organs are frozen by structures like morgue trays, freezer closets and freezer crates
add: Can't do surgeries on frozen bodies or when the surgery is targeting a specific organ and that organ is frozen

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Organs now freeze at 0 celsius (including organs in dead bodies)

- Frozen organs now look visibly frozen to indicate to the player that they're preserved
- Humans with body temperatures of 0 celsius and lower now look paler to indicate that their organs are preserved if they're dead

Full details:

obj_flags.dm
- ORGAN_FROZEN organ_flag is now ORGAN_NO_DECAY, for use in cases where organs shouldn't decay but shouldn't freeze either

traits.dm
- Adds two new traits, TRAIT_ON_ICE, which indicates an object or mob should begin freezing or stay frozen, and TRAIT_COLD_SKIN, which indicates a human's skintone has changed to be lighter from cold

game.dm
- Instead of inverting the ORGAN_FROZEN flag, recursive_organ_check now adds or removes the TRAIT_ON_ICE trait to carbons and organs, depending on whether they already have it and what it was added by
- Carbons now receive the TRAIT_ON_ICE trait under recursive_organ_check(), which tries to freeze their organs when added

icons.dm
- Adds begin_freeze() and begin_unfreeze() signal handlers at atom level, which are called when an organ or mob receives the TRAIT_ON_ICE trait, and call make_frozen() and make_unfrozen() after three seconds
- Adds the make_frozen() and make_unfrozen() procs at atom level 
- Overrides make_frozen() proc for carbons, which gives organs inside them the on ice trait and immediately freezes their organs
- Adds the make_frozen() proc for humans, which adds the cold skin trait and updates their body parts to update their skintone to a lighter shade
- Adds make_unfrozen() procs for carbons and humans, which undo the effects of the procs above
- Renames make_frozen_visual() object proc to make_frozen(), as in addition to updating object visuals it adds the FROZEN obj_flag

traits.dm
- Adds TRAIT_ON_ICE and TRAIT_COLD_SKIN to list of mob traits

medkit.dm
- Uses TRAIT_ON_ICE instead of ORGAN_FROZEN organ flag

_open.dm
- Uses TRAIT_ON_ICE in addition to renamed make_frozen() proc to immediately freeze objects, then removes the trait so they can be thawed when entering a temperate turf

turf.dm
- Makes it so frozen objects entering temperate turfs call begin_unfreeze() so they take 3 seconds to thaw out, for consistency 

smartfridge.dm
- Uses on ice trait instead of ORGAN_FROZEN flag

mapping_helpers.dm
- Uses on ice trait for the dead human instead of adding ORGAN_FROZEN to their organs

tendril_loot.dm
- Uses ORGAN_NO_DECAY instead of ORGAN_FROZEN on dark shard to prevent it decaying before being used without freezing them

MMI.dm
- Uses ORGAN_NO_DECAY instead of ORGAN_FROZEN on brains in MMI to prevent them decaying in the MMI without freezing them

carbon.dm
- Register signal in initialize for TRAIT_ON_ICE being added or removed to call begin_freeze() or begin_unfreeze() on carbons

examine.dm
- TRAIT_COLD_SKIN gives the same message as albinism

human_update_icons.dm
- TRAIT_COLD_SKIN adds to icon_render_key so icon is updated when it's added

species.dm
- TRAIT_ON_ICE is added when the human's body temperature is below 0 celsius, and removed when it's above that

robot.dm
- Brain in MMI uses ORGAN_NO_DECAY

impure_medicine_reagents.dm
- Original heart in corazargh code uses ORGAN_NO_DECAY

biodome_winter.dm
- Freeze cube freezes objects immediately, then removes the TRAIT_ON_ICE trait so they can be thawed when entering a temperate turf

_bodyparts.dm
- If the owner pf the body part is human and they have cold skin, ligthen the color of it slightly

organ_internal.dm
- Register signals in organ Initialize() on adding and removing TRAIT_ON_ICE to try to freeze or unfreeze the organ
- Add TRAIT_ON_ICE if we're on a turf beneath 0 celsius, otherwise remove it if we have it from a frozen turf
- Do not apply organ damage to a dead organ if it has ORGAN_NO_DECAY or the FROZEN obj_flag

cacogen committed:
Can't operate on frozen bodies or frozen organs

GuillaumePrata commented:
Plasmamen's base body temperature is -3C, how would it work for them?

cacogen committed:
Frozen organs in living mobs take damage

- Use p_are() to refer to organ to manipulate
- Add visible message for freezing humans, as the change in skintone is subtle and may look like a bug to some people

SuperNovaa41 commented:
> Can't do surgeries with a set organ to manipulate if that organ is frozen or do surgeries on a dead body if it's frozen

This isn't mentioned in your changelog and its arguably the most important thing to be mentioned

cacogen committed:
Gets rid of visible messages on going pale (didn't like them)

- Makes skintone turn a sickly blue colour when too cold

cacogen commented:
> Plasmamen's base body temperature is -3C, how would it work for them?

Thanks for bringing this to my attention. 

It doesn't cause their organs to freeze or prevent surgery, because these only happen when the mob is dead. 

When dead, their bodies are no longer pulling them down to -3 celsius so you'll be able to do surgeries on them.  

However, they do say they have pale skin when examined, which is something I'll have to fix. 

This also made me realise there's nothing to make organs freeze on death if the body is already freezing, so that's something I'll have to fix too.

cacogen committed:
Organs are frozen in living humans if their core temperature is beneath freezing and their species' cold damage limit

- Removes reference to skin in human pale examine message for species without skin
- Stops the pale examine message appearing for mobs that don't use skin but have skintone set to albino
- Lightens skin slightly more on freezing to make it more obvious

cacogen committed:
Don't have to be dead to be frozen (but it helps)

cacogen commented:
I've changed it and now the pale skin message doesn't reference skin (plasmamen don't have skin), and it only shows up when the human mob is beneath their species' cold damage limit (i.e. is taking cold damage). This means plasmamen won't have the message by default, due to having core temperatures beneath 0 celsius.

I've also made it so a body can freeze while alive (making it pale and freezing its organs and preventing surgery on it), but only if its core temperature is beneath 0 celsius and its species' cold damage limit. This means you will be able to do surgery on plasmamen as usual and they won't be walking around with frozen organs by default.

cacogen committed:
I've made a terrible mistake

cacogen committed:
Frozen organs are failing organs

cacogen commented:
I added preventing surgery on frozen bodies because when I saw Cobby assigned, based on forum threads about advanced surgeries and stasis beds they've posted in, I figured they'd be against having an easy way to do advanced surgeries without having to worry about organ decay. 

It's fairly easy to make a small room cold enough to freeze corpses with this (particularly on Ice Box, it's not like corpses breathe), while you wear a winter coat to protect you against damage.

>does having an immunity to cold (such as from, say, the temperature immunity mutation) make you unaffected by this? what about being in a cryo tube? a stasis bed?

Temperature adaptation gives you the RESIST_COLD trait which prevents the block of code in body_temperature_damage() from running where the ON_ICE trait is added (which freezes your organs) if your core temperature is low enough.

Cryo tubes stop body temperature from regulating altogether, so organs won't freeze if you're in one. They won't unfreeze either, which I'll have to figure out what to do about.

Stasis beds stop body_temperature_damage() from running when you're in one, so organs won't freeze. They won't unfreeze either, which I'll have to figure out what to do about.

>will temp guns, cryokinesis, and/or changeling cryo stings become incredibly lethal because of this? what's this threshold's relation to the cryo pistol's victim detonation threshold?

I tested the changeling cryo sting, and if you stacked it about 4 times on a human their organs failed after about a minute and a half and by the two minute mark they were dead. 

That made me revert frozen organs being failing organs and just make it so that frozen organs take organ damage of the same amount as EMP'd cybernetic organs (which is an amount that could be changed based on balance needs).

I'll have to test temperature guns and cryokinesis to see what their effects are.

Thank you for this feedback.

>if you really, really think that using a cold room instead of formaldehyde to preserve corpses you're performing surgery on is a gamebreaking issue, you can just multiply surgery lengths by 1.5x (or whatever penalty cryostylane uses) if the subject is cold instead of blocking them altogether.

I could look into doing that instead.

cacogen committed:
Merge remote-tracking branch 'origin/master' into organfreeze

ShadowAKT commented:
How would this work with cybernetic organs? It would make sense for the tiers above basic to be immune to cold, at the very least.

What if you were to transplant a plasmaman's organ into a human? Would the organ lose its cold immunity when it wouldn't make sense to?

Also would cryo cause organ death, since it lowers body temp? I would imagine you'd have to change cryoxadone to prevent this.

cacogen commented:
>How would this work with cybernetic organs? It would make sense for the tiers above basic to be immune to cold, at the very least.

They freeze like regular organs in certain cases, but not others. They don't freeze on cold turfs, but they do freeze if they're inside a body in a morgue tray, for example. I'll make it consistent.

>What if you were to transplant a plasmaman's organ into a human? Would the organ lose its cold immunity when it wouldn't make sense to?

Their organs don't have cold immunity. Organs freeze in human mobs when their core temperature is both below 0 celsius and below that mob's threshold for taking cold damage. Plasmamen and humans have different thresholds for taking cold damage. Plasmamen's is much lower than humans'.

>Also would cryo cause organ death, since it lowers body temp?

The organs can't freeze because a cryo check stops natural body temperature regulation, which prevents the bit of code that causes cold damage from running (which is where the code that causes organs to freeze if the mob's core temperature is below the cold damage threshold is).

Kylerace commented on code/__HELPERS/icons.dm:
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
+		return
+	addtimer(CALLBACK(src, .proc/make_unfrozen), 3 SECONDS)
+
+/atom/proc/make_frozen()
```
doc

Kylerace commented on code/__HELPERS/icons.dm:
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
+	SIGNAL_HANDLER
+	addtimer(CALLBACK(src, .proc/make_frozen), 3 SECONDS)
+
+/atom/proc/begin_unfreeze()
```
doc

Kylerace commented on code/__HELPERS/icons.dm:
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
 
-/obj/proc/make_frozen_visual()
-	// Used to make the frozen item visuals for Freon.
+/atom/proc/begin_freeze()
```
doc

Kylerace replied:
also all of these should go into their type paths relevant file like the atom ones to the atom level and the carbon ones to the carbon level

Kylerace replied:
also why are these on /atom level if theyre only used by /atom/movable

Kylerace commented on code/__HELPERS/icons.dm:
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
+/atom/proc/make_frozen()
+	return
+
+/atom/proc/make_unfrozen()
```
doc

Kylerace commented on code/__HELPERS/icons.dm:
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
+/mob/living/carbon/human/make_frozen()
+	if(!HAS_TRAIT(src, TRAIT_ON_ICE))
+		return
+	if(HAS_TRAIT_FROM(src, TRAIT_COLD_SKIN, "make_frozen()")) //already have it, don't need to apply it again
```
dont name trait sources after functions like this its hard to work with. even if that source is unique to that proc it should be something like "frozen solid" or whatever

Kylerace commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1672,6 +1673,19 @@ GLOBAL_LIST_EMPTY(features_by_species)
+	if(humi.bodytemperature < cold_damage_limit && !HAS_TRAIT(humi, TRAIT_RESISTCOLD))
+		if(HAS_TRAIT_FROM(humi, TRAIT_COLD_SKIN, "body temp"))
+			return
+		ADD_TRAIT(humi, TRAIT_COLD_SKIN, "body temp")
```
trait sources like these should be defines

Kylerace commented on code/modules/ruins/lavalandruin_code/biodome_winter.dm:
```diff
@@ -42,7 +42,9 @@
 		if(!(hit_object.obj_flags & FROZEN))
-			hit_object.make_frozen_visual()
+			ADD_TRAIT(hit_object, TRAIT_ON_ICE, type)
+			hit_object.make_frozen() //immediately freeze
```
why do you need to add a trait to an object just to call a proc and then remove the trait

cacogen replied:
I don't like it either but the proc doesn't work without the trait. How would you do it?

Kylerace commented on code/modules/surgery/organs/organ_internal.dm:
```diff
@@ -51,6 +51,9 @@ INITIALIZE_IMMEDIATE(/obj/item/organ)
 			volume = reagent_vol,\
 			after_eat = CALLBACK(src, .proc/OnEatFrom))
+
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_ON_ICE), /atom.proc/begin_freeze)
```
this is a bad design pattern its confusing as to what the future programmer who has to work with your code is doing. youre essentially using adding this trait to an organ is a proc call which isnt clear at all unless someone finds these two lines.

cacogen replied:
So, remove the signal and call the proc directly?

Kylerace reviewed (changes requested):
im wary of this. not only are you adding to two parts of the Life() loop youre attaching this behavior to every organ in a carbon and im not sure what the benefits over the frozen organ flag this has

cacogen commented:
Is the Life() loop not the right place to do it?

I made it a trait because there can be multiple things freezing an organ simultaneously and I needed to keep track of those things to prevent organs unfreezing prematurely when one is removed.

By making it use the obj_flag it means that other sources of freezing also prevent an organ's decay, which is important for consistency when making frozen organs look like other frozen objects.

cacogen commented:
It was not a good idea to have freezing organs fail, because it made things that lowered body temperature drastically like the cryo sting ridiculously OP. Instead now they just take damage like EMP'd cybernetic ones.

It's quite difficult to get your core temperature that low (at least as a human). When I tested this I needed to be in like a -50 celsius room without any clothes for it to happen.

LemonInTheDark commented:
I keep meaning to fix that
